--without-zlib does not work

Using greenplum 6X-STABLE branch:

```
$ ./configure --without-zlib
$ make
gdb -O0 -g3 -ggdb3 -std=gnu99   -Werror=uninitialized -Werror=implicit-function-declaration -I../../../src/include -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk -I/usr/local/include/ -I/usr/local/opt/openssl/include -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/libxml2   -c -o pg_compression.o pg_compression.c
pg_compression.c:60:22: error: unknown type name 'Bytef'
        int (*compress_fn) (Bytef *dest,
                            ^
pg_compression.c:61:7: error: unknown type name 'uLongf'
                                                uLongf *destLen,
                                                ^
pg_compression.c:62:13: error: unknown type name 'Bytef'
                                                const Bytef *source,
                                                      ^
pg_compression.c:63:7: error: unknown type name 'uLong'
                                                uLong sourceLen,
                                                ^
pg_compression.c:66:24: error: unknown type name 'Bytef'
        int (*decompress_fn) (Bytef *dest,
                              ^
pg_compression.c:67:9: error: unknown type name 'uLongf'
                                                  uLongf *destLen,
                                                  ^
pg_compression.c:68:15: error: unknown type name 'Bytef'
                                                  const Bytef *source,
                                                        ^
pg_compression.c:69:9: error: unknown type name 'uLong'
                                                  uLong sourceLen);
                                                  ^
pg_compression.c:249:23: error: use of undeclared identifier 'compress2'; did you mean
      'compress'?
        state->compress_fn = compress2;
                             ^~~~~~~~~
                             compress
pg_compression.c:237:11: note: 'compress' declared here
        bool                      compress = PG_GETARG_BOOL(2);
                                  ^
pg_compression.c:249:21: warning: incompatible integer to pointer conversion assigning to 'int
      (*)(int *, int *, const int *, int, int)' from 'bool' (aka 'char') [-Wint-conversion]
        state->compress_fn = compress2;
                           ^ ~~~~~~~~~
pg_compression.c:250:25: error: use of undeclared identifier 'uncompress'; did you mean
      'compress'?
        state->decompress_fn = uncompress;
                               ^~~~~~~~~~
                               compress
pg_compression.c:237:11: note: 'compress' declared here
        bool                      compress = PG_GETARG_BOOL(2);
                                  ^
pg_compression.c:250:23: warning: incompatible integer to pointer conversion assigning to 'int
      (*)(int *, int *, const int *, int)' from 'bool' (aka 'char') [-Wint-conversion]
        state->decompress_fn = uncompress;
                             ^ ~~~~~~~~~~
pg_compression.c:283:34: warning: incompatible pointer types passing 'unsigned char *' to
      parameter of type 'int *' [-Wincompatible-pointer-types]
        last_error = state->compress_fn((unsigned char *) dst,
                                        ^~~~~~~~~~~~~~~~~~~~~
pg_compression.c:284:10: warning: incompatible pointer types passing 'unsigned long *' to
      parameter of type 'int *' [-Wincompatible-pointer-types]
  ...&amount_available_used, src, src_sz,
     ^~~~~~~~~~~~~~~~~~~~~~
pg_compression.c:289:20: error: use of undeclared identifier 'Z_OK'
        if (last_error != Z_OK)
                          ^
pg_compression.c:293:9: error: use of undeclared identifier 'Z_MEM_ERROR'
                        case Z_MEM_ERROR:
                             ^
pg_compression.c:297:9: error: use of undeclared identifier 'Z_BUF_ERROR'
                        case Z_BUF_ERROR:
                             ^
pg_compression.c:335:19: error: unknown type name 'Bytef'
                                                                          (const Byte...
                                                                                 ^
pg_compression.c:339:20: error: use of undeclared identifier 'Z_OK'
        if (last_error != Z_OK)
                          ^
pg_compression.c:343:9: error: use of undeclared identifier 'Z_MEM_ERROR'
                        case Z_MEM_ERROR:
                             ^
pg_compression.c:347:9: error: use of undeclared identifier 'Z_BUF_ERROR'
                        case Z_BUF_ERROR:
                             ^
pg_compression.c:357:9: error: use of undeclared identifier 'Z_DATA_ERROR'
                        case Z_DATA_ERROR:
                             ^
4 warnings and 18 errors generated.
make[2]: *** [pg_compression.o] Error 1
make[1]: *** [catalog-recursive] Error 2
make: *** [all-backend-recurse] Error 2
```

--- a/src/backend/access/common/reloptions_gp.c
+++ b/src/backend/access/common/reloptions_gp.c
@@ -930,16 +930,24 @@ validate_and_adjust_options(StdRdOptions *result,
 		/* Check upper bound of compresslevel for each compression type */
 
 		if (result->compresstype[0] &&
-			(pg_strcasecmp(result->compresstype, "zlib") == 0) &&
-			(result->compresslevel > 9))
+			(pg_strcasecmp(result->compresstype, "zlib") == 0))
 		{
-			if (validate)
-				ereport(ERROR,
-						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
-						 errmsg("compresslevel=%d is out of range for zlib (should be in the range 1 to 9)",
-								result->compresslevel)));
+#ifndef HAVE_LIBZ
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("zlib compression is not supported by this build"),
+					 errhint("Compile without --without-zlib to use zlib compression.")));
+#endif
+			if (result->compresslevel > 9)
+			{
+				if (validate)
+					ereport(ERROR,
+							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+							 errmsg("compresslevel=%d is out of range for zlib (should be in the range 1 to 9)",
+									result->compresslevel)));
 
-			result->compresslevel = setDefaultCompressionLevel(result->compresstype);
+				result->compresslevel = setDefaultCompressionLevel(result->compresstype);
+			}
 		}
 
 		if (result->compresstype[0] &&
@@ -1169,13 +1177,21 @@ validateAppendOnlyRelOptions(bool ao,
 					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
 					 errmsg("compresstype cannot be used with compresslevel 0")));
 
-		if (comptype && (pg_strcasecmp(comptype, "zlib") == 0) &&
-			(complevel < 0 || complevel > 9))
+		if (comptype && (pg_strcasecmp(comptype, "zlib") == 0))
 		{
+#ifndef HAVE_LIBZ
 			ereport(ERROR,
-					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
-					 errmsg("compresslevel=%d is out of range (should be between 0 and 9)",
-							complevel)));
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("zlib compression is not supported by this build"),
+					 errhint("Compile without --without-zlib to use zlib compression.")));
+#endif
+			if (complevel < 0 || complevel > 9)
+			{
+				ereport(ERROR,
+						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						 errmsg("compresslevel=%d is out of range (should be between 0 and 9)",
+								complevel)));
+			}
 		}
 
 		if (comptype && (pg_strcasecmp(comptype, "zstd") == 0))

--- a/src/backend/catalog/pg_compression.c
+++ b/src/backend/catalog/pg_compression.c
@@ -48,6 +48,7 @@ char *storage_directive_names[] = {"compresstype", "compresslevel",
 								   "blocksize", NULL};
 
 
+#ifdef HAVE_LIBZ
 /* Internal state for zlib */
 typedef struct zlib_state
 {
@@ -69,6 +70,7 @@ typedef struct zlib_state
 						  uLong sourceLen);
 
 } zlib_state;
+#endif
 
 static NameData
 comptype_to_name(char *comptype)
@@ -224,6 +226,7 @@ callCompressionValidator(PGFunction func, char *comptype, int32 complevel,
 	(void)DirectFunctionCall1(func, PointerGetDatum(&sa));
 }
 
+#ifdef HAVE_LIBZ
 Datum
 zlib_constructor(PG_FUNCTION_ARGS)
 {
@@ -378,6 +381,42 @@ zlib_validator(PG_FUNCTION_ARGS)
 {
 	PG_RETURN_VOID();
 }
+#else
+Datum
+zlib_constructor(PG_FUNCTION_ARGS)
+{
+	elog(ERROR, "libz compression is not supported in this build of Greenplum");
+	PG_RETURN_VOID();
+}
+
+Datum
+zlib_destructor(PG_FUNCTION_ARGS)
+{
+	elog(ERROR, "libz compression is not supported in this build of Greenplum");
+	PG_RETURN_VOID();
+}
+
+Datum
+zlib_compress(PG_FUNCTION_ARGS)
+{
+	elog(ERROR, "libz compression is not supported in this build of Greenplum");
+	PG_RETURN_VOID();
+}
+
+Datum
+zlib_decompress(PG_FUNCTION_ARGS)
+{
+	elog(ERROR, "libz compression is not supported in this build of Greenplum");
+	PG_RETURN_VOID();
+}
+
+Datum
+zlib_validator(PG_FUNCTION_ARGS)
+{
+	elog(ERROR, "libz compression is not supported in this build of Greenplum");
+	PG_RETURN_VOID();
+}
+#endif
 
 Datum
 rle_type_constructor(PG_FUNCTION_ARGS)
@@ -456,9 +495,6 @@ dummy_compression_validator(PG_FUNCTION_ARGS)
 bool
 compresstype_is_valid(char *comptype)
 {
-	bool found = false;
-	int i;
-
 	/*
 	 * Hard-coding compresstypes is bad, agreed.  But there isn't a
 	 * better way in sight at this point.  Lookup into pg_compression
@@ -470,14 +506,25 @@ compresstype_is_valid(char *comptype)
 	 * Whenever the list of supported compresstypes is changed, this
 	 * must change!
 	 */
-	static const char *const valid_comptypes[] =
-			{"quicklz", "zlib", "rle_type", "none", "zstd"};
-	for (i = 0; !found && i < ARRAY_SIZE(valid_comptypes); ++i)
+	static const char *const valid_comptypes[] = {
+#ifdef HAVE_LIBQUICKLZ
+			"quicklz",
+#endif
+#ifdef HAVE_LIBZ
+			"zlib",
+#endif
+#ifdef HAVE_LIBZSTD
+			"zstd",
+#endif
+			"rle_type", "none"};
+
+	for (int i = 0; i < ARRAY_SIZE(valid_comptypes); ++i)
 	{
 		if (pg_strcasecmp(valid_comptypes[i], comptype) == 0)
-			found = true;
+			return true;
 	}
-	return found;
+
+	return false;
 }
 
 /*

--- a/src/backend/utils/misc/fstream/gfile.c
+++ b/src/backend/utils/misc/fstream/gfile.c
@@ -32,6 +32,7 @@
 #include <fcntl.h>
 #include <sys/stat.h>
 #include <sys/file.h>   /* for flock */
+#include <unistd.h>
 
 #ifdef WIN32
 #include <io.h>

--- a/src/include/access/reloptions.h
+++ b/src/include/access/reloptions.h
@@ -34,7 +34,11 @@
  * If compression is turned on without specifying compresstype, this
  * is the default.
  */
+#ifdef HAVE_LIBZ
 #define AO_DEFAULT_COMPRESSTYPE   "zlib"
+#else
+#define AO_DEFAULT_COMPRESSTYPE   "none"
+#endif
 #define AO_DEFAULT_CHECKSUM       true
 #define AO_DEFAULT_COLUMNSTORE    false
 

